core/src/main/scala/kafka/log/Log.scala

```diff
@@ -399,11 +400,11 @@ class Log(@volatile var dir: File,
           case _: NoSuchFileException =>
             error(s"Could not find offset index file corresponding to log file ${segment.log.file.getAbsolutePath}, " +
               "recovering segment and rebuilding index files...")
-            recoverSegment(segment)
+            segment.recover(new ProducerStateManager(topicPartition, dir, maxProducerIdExpirationMs))
```

Comment:
We talked extensively about this offline but I am trying to summarize some of the important points.

- It is quite odd (to me) that we care so much about a "corrupt" index, given that we have minimal sanity checks on the log itself. The common case I can think of hitting this case is when partial index data is written to disk (because of unclean shutdown, for example), or when indexes do not exist at all (for example when user is upgrading from a version of Kafka that does have time indexes to a version that does). If these are the most likely cases, I would much rather call into `segment.createIndices` here instead of `segment.recover` to make this distinction clear but this probably calls for another PR so that we can have a more focussed discussion about this.

- We maintain producer snapshots for the last two segments after a checkpoint is written. This means the invariant on startup is that we always have at least two snapshot files from when the last checkpoint was taken. All segments from before the time of the checkpoint remain intact and we do not attempt any "recovery" on these. To build up the full producer state, we need to read log data from this last snapshot till the end of the log, in the worst case. Ideally, we should never have to read the log from the beginning to build up the producer state.

- If we do find a real corruption in log data, I think we should take the log directory offline. If all log directories are offline, we should take the broker down. I am not sure if we already do this but I can check to verify. The rationale is that if we find a corruption, we do not know what to trust and what not to, and it probably is not safe to continue operating at this point.

Reply:
I think basically what we agreed offline is that we will try to decouple producer state loading from segment recovery. The current logic is pessimistic. We assume that failing a sanity check or finding a missing index means we'll need to load producer state for the corresponding segment. Instead, we'll try to be more optimistic. Assume there is nothing wrong with the segment and skip producer state loading if we know we have a snapshot at a higher offset. Does that sound right?

Reply:
Yeah, that sounds right. Another way to put this is that we must only read through the segments after the recovery point to build up the producer state.

Reply:
Unfortunately, this optimization is kind of tricky. There is a case for reading through the log data if the transaction index for a particular segment is either missing, or is corrupt. Because this needs more thought, I think if we tackle this problem separately. For now, I reverted to the old behavior where we have to read from the last snapshot when we find a corrupted index.